File: Python/002-AddTwoNumbers/SolutionByRecursion.py
# !/usr/bin/env Python3
# -*- coding: utf-8 -*-

#  @Author: Frenude 
#  @Description:
#  @Date: Created in 01 05,2021
#  @Modified By:

class ListNode():
    def __init__(self, val, next=None):
        if isinstance(val, int):
            self.val = val
            self.next = next

        elif isinstance(val, list):
            self.val = val[0]
            self.next = next
            cur = self
            for i in val[1:]:
                cur.next = ListNode(i)
                cur = cur.next

    def gatherAttrs(self):
        return ", ".join("{}: {}".format(k, getattr(self, k)) for k in self.__dict__.keys())

    def __str__(self):
        return self.__class__.__name__ + " {" + "{}".format(self.gatherAttrs()) + "}"


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        def addTwoNumber(l1: ListNode, l2: ListNode, carry: int) -> ListNode:
            if not l1 and not l2 and not carry: return None
            sum = (l1.val if l1 else 0) + (l2.val if l2 else 0) + carry
            node = ListNode(sum % 10)
            node.next = addTwoNumber(l1.next if l1 else None, l2.next if l2 else None, sum // 10)
            return node
        return  addTwoNumber(l1,l2,0)


if __name__ == '__main__':
    l1 = ListNode([9, 9, 9, 9, 9, 9, 9])
    l2 = ListNode([9, 9, 9, 9])
    s = Solution()
    ret = s.addTwoNumbers(l1, l2)
    print(ret)

File: Java/002-AddTwoNumbers/SolutionByRecursion.java
/**
 * @Author: Frenude
 * @Description: //TODO
 * @Date: Created in 01 05,2021
 * @Modified By:
 **/


public class SolutionByRecursion {

    public static void main(String[] args) {
        ListNode l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
        ListNode l2 = new ListNode(5, new ListNode(6, new ListNode(4)));

        SolutionByRecursion solution = new SolutionByRecursion();
        System.out.println(solution.addTwoNumbers(l1, l2));
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode listNode = addTwoNumber(l1, l2, 0);
        return listNode;
    }
    public ListNode addTwoNumber(ListNode l1, ListNode l2 ,int carry){
        if (l1 == null && l2 == null && carry == 0) return null;
        int sum = (l1 != null ? l1.val : 0) + (l2 != null ? l2.val : 0) + carry;
        ListNode node = new ListNode(sum % 10);
        node.next = addTwoNumber(l1 != null ? l1.next : null, l2 != null ? l2.next : null, sum / 10);
        return node;
    }
}
